File: 2023/Leetcode/problem102.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from queue import Queue

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        
        result = [[]]
        queue = Queue()
        queue.put(root)
        currentDepth = 0
        depths = {root: 0}
        
        while not queue.empty():
            root = queue.get()

            if root:
                if depths[root] != currentDepth:
                    result.append([])
                    currentDepth = depths[root]

                result[-1].append(root.val)
                    
                queue.put(root.left)
                depths[root.left] = currentDepth + 1

                queue.put(root.right)
                depths[root.right] = currentDepth + 1

        return result

File: 2024/Leetcode/problem102.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> result;

        if (!root) {
            return result;
        }

        queue<size_t> levels;
        queue<TreeNode*> myQueue;
        
        myQueue.push(root);
        levels.push(0);

        while (!myQueue.empty()) {
            TreeNode* currNode = myQueue.front();
            auto currLevel = levels.front();

            myQueue.pop();
            levels.pop();

            if (result.size() == currLevel) {
                result.push_back(vector<int>());
            }
            result[currLevel].push_back(currNode -> val);

            if (currNode -> left) {
                myQueue.push(currNode -> left);
                levels.push(currLevel + 1);
            }

            if (currNode -> right) {
                myQueue.push(currNode -> right);
                levels.push(currLevel + 1);
            }
        }

        return result;
    }
};